src/vaults/SectorCrossVault.sol

```diff
@@ -0,0 +1,344 @@
+	enum msgType {
+		NONE,
+		DEPOSIT,
+		REDEEM,
+		REQUESTREDEEM,
+		REQUESTVALUEOFSHARES,
+		EMERGENCYWITHDRAW
+	}
+
+	struct Vault {
+		uint16 chainId;
+		address adapter;
+		bool allowed;
+	}
+
+	struct Request {
+		uint256 timestamp;
+		uint256 chainId;
+		address vault;
+	}
+
+	struct HarvestLedger {
+		uint256 depositValue;
+		bool isOpen;
+		uint256 openIndex;
+		Request[] request;
+	}
+
+	// TODO Implement functions with harvestLock modifier
+
+	// Controls deposits
+	mapping(address => Vault) public depositedVaults;
+
+	// Keep track of depositedVaults
+	// TODO Has to implement push to this array
+	address[] internal vaultsArr;
+
+	// Harvest state
+	HarvestLedger public harvestLedger;
+	// Controls emergency withdraw
+	bool internal emergencyEnabled = false;
+
+	constructor(
+		ERC20 _asset,
+		string memory _name,
+		string memory _symbol,
+		address _owner,
+		address _guardian,
+		address _manager,
+		address _treasury,
+		uint256 _perforamanceFee
+	) ERC4626(_asset, _name, _symbol, _owner, _guardian, _manager, _treasury, _perforamanceFee) {
+		// Not sure if needed
+		// TODO check if needed
+		harvestLedger.openIndex = 0;
+		harvestLedger.isOpen = false;
+		harvestLedger.depositValue = 0;
+	}
+
+	/*/////////////////////////////////////////////////////
+					Cross Vault Interface
+	/////////////////////////////////////////////////////*/
+
+	function depositIntoVaults(address[] calldata vaults, uint256[] calldata amounts)
+		public
+		onlyRole(MANAGER)
+		checkInputSize(vaults.length, amounts.length)
+	{
+		for (uint256 i = 0; i < vaults.length; ) {
+			Vault memory tmpVault = depositedVaults[vaults[i]];
+
+			if (!tmpVault.allowed) revert VaultNotAllowed(vaults[i]);
+
+			if (tmpVault.adapter == address(0)) {
+				BatchedWithdraw(vaults[i]).deposit(amounts[i], address(this));
+			} else {
+				IXAdapter(tmpVault.adapter).sendMessage(
+					amounts[i],
+					vaults[i],
+					address(this),
+					tmpVault.chainId,
+					uint16(msgType.DEPOSIT),
+					uint16(block.chainid)
+				);
+
+				emit BridgeAsset(uint16(block.chainid), tmpVault.chainId, amounts[i]);
+			}
+
+			unchecked {
+				i++;
+			}
+		}
+	}
+
+	function requestRedeemFromVaults(address[] calldata vaults, uint256[] calldata shares)
+		public
+		onlyRole(MANAGER)
+		checkInputSize(vaults.length, shares.length)
+	{
+		for (uint256 i = 0; i < vaults.length; ) {
+			Vault memory tmpVault = depositedVaults[vaults[i]];
+
+			if (!tmpVault.allowed) revert VaultNotAllowed(vaults[i]);
+
+			if (tmpVault.adapter == address(0)) {
+				BatchedWithdraw(vaults[i]).requestRedeem(shares[i]);
+			} else {
+				IXAdapter(tmpVault.adapter).sendMessage(
+					shares[i],
+					vaults[i],
+					address(this),
+					tmpVault.chainId,
+					uint16(msgType.REQUESTREDEEM),
+					uint16(block.chainid)
+				);
+			}
+
+			unchecked {
+				i++;
+			}
+		}
+	}
+
+	function redeemFromVaults(address[] calldata vaults, uint256[] calldata shares)
```

Comment:
would it be better to iterate through the x-vault requests on the c-vault and send funds back during the withdrawal process? this can save us the extra cross chain message.

Reply:
In this case I'll remove this function completely, right?